### DISCUSSION TITLE
command output file expansion

### DISCUSSION THREAD

**Discussion by kenrector** [ImageMagick/ImageMagick#1942] [General] [answered]
I'm having trouble producing multiple output files for a convert command with *.png for input.  I'm using Mac OS 10.15.4 (Catalina) with zsh.  IM version:
Version: ImageMagick 7.0.10-7 Q16 x86_64 2020-04-10 https://imagemagick.org

The command is:

convert *.png \( -clone 0 -morphology close rectangle:40x1 -negate \) -compose plus -composite img%d.png   

The result is just one file although the command takes a long time and I feel its converting all five image*.png files.  Also, the resultant file (img0.png)  is garbage that looks like all outputs have been combined together.  Executing the convert with  unparameterized file names produces a correct output for that single file.  I believe I've done this before and it worked ok, but I can't get it to work now.  

Can someone see what I've done wrong?

An ls after the convert:
sds/ocropus_symbol/cropped % ls -l
total 912
-rw-r--r--@ 1 admin  staff   47435 Apr 24 21:20 image001.png
-rw-r--r--@ 1 admin  staff   63792 Apr 24 21:20 image002.png
-rw-r--r--@ 1 admin  staff  122556 Apr 24 21:21 image003.png
-rw-r--r--  1 admin  staff   62475 Apr 24 21:20 image004.png
-rw-r--r--@ 1 admin  staff  127579 Apr 24 21:21 image005.png
-rw-r--r--  1 admin  staff   32128 Apr 26 16:32 img0.png

and a snippet from img0.png:

![Screen Shot 2020-04-26 at 4 59 29 PM](https://user-images.githubusercontent.com/9969282/80323414-1603e700-87e0-11ea-937b-9c97950d717c.png)

and a snippet convert image001.png:

![Screen Shot 2020-04-26 at 5 21 58 PM](https://user-images.githubusercontent.com/9969282/80323832-94618880-87e2-11ea-8383-9fc3b936e707.png)

Further, I'm having trouble expanding a shell variable for the file name when the command is used in a zsh script.  But that's maybe a story for a different occasion.

**Comment by fmw42** [comment, marked as answer]
To composite one image onto a multiple of images, you need to use -layers composite construction and separate the two groups by null:

`convert *.png null: ( -clone 0 -morphology close rectangle:40x1 -negate ) -compose plus -layers composite img%d.png
`

Assuming that is what you want, see https://imagemagick.org/Usage/anim_mods/#composite

I am not sure that is what you want. But in general, processing multiple inputs using wild cards and having multiple outputs is not done with convert, but with mogrify or in some case, you can do that in ImageMagick 7 with inline argument computations.

**Reply by kenrector** [reply]
Thank you Fred,
What I wanted was to simply batch process a bunch of images and save the output files.  I looked at mogrify but it doesn't seem to handle -clone and I wanted to avoid learning about that.  

My attempt at a shell script didn't work, and I couldn't figure that out

I thought that the imagemagick wild card input would do what I want but I was mistaken.

Now I see that the problem in my shell script has something to do with how it's invoked, as in  ".  close.sh" as opposed to "close.sh".  I've forgotten what that shell sourcing is all about, but I'll figure that out.

So, sorry to have bothered you.

**Reply by fmw42** [reply]
Not a bother. Was my result what you had wanted? If not explain further what it is you want to happen.

**Reply by kenrector** [reply]
I thought I explained above.  It's not about compositing multiple images.  Its simpler than that.

I just want to batch process a bunch of files and save the individual resultant files.

**Comment by fmw42** [comment]
If you need to use parenthesis processing, you can use a bash "for" loop as follows. Assume jpg images are in folder, test1 that you want to process and you want the results in an existing folder, test2 at the same level in your directory structure. Then

```
cd
cd desktop/test1
for img in *.jpg; do
convert "$img" \( -clone 0 -morphology close rectangle:40x1 -negate \) -compose plus -composite "../test2/$img"
done
```

If you do not need parenthesis processing, then use mogrify.

**Reply by kenrector** [reply]
Yes, see above, the scripting all works ok and I'm good to go.  However, while I have your attention;

With your help in post #1916 I nearly eliminated the horizontal lines from the image.
The morphology close operation leaves a one pixel wide outline around the area.   Here's a snippet.
![image000](https://user-images.githubusercontent.com/9969282/80399380-f3b8aa80-886d-11ea-829d-ef90b58fcd83.png)

I assume that's a result of how the close (dilate, erode) operations work.  I tried different size kernels and other experiments (guesswork)  but I didn't find a way to get rid of these.  I didn't try higher level stuff like despeckel because I didn't want to affect the text characters.

Do you think there is a way to get rid of these outlines?

**Reply by fmw42** [reply]
You now can try using a smaller kernel in morphology close

`convert s1.png -morphology close diamond:1 x.png`

![x](https://user-images.githubusercontent.com/13458218/80402442-c6223000-8872-11ea-922c-9b01cb0cca2c.png)
